File: backends/fastapi/app/usecase/events.py
from __future__ import annotations

from dataclasses import dataclass

from app.domain import errors
from app.domain.models import Event, Sector
from app.usecase.ports import EventRepository, SectorRepository


@dataclass
class EventDetail:
    event: Event
    sectors: list[Sector]


class EventService:
    def __init__(self, events: EventRepository, sectors: SectorRepository):
        self._events = events
        self._sectors = sectors

    async def list(self, cursor: str, limit: int) -> tuple[list[Event], str]:
        if limit <= 0 or limit > 100:
            limit = 20
        return await self._events.list(cursor, limit)

    async def get(self, event_id: str) -> EventDetail:
        event = await self._events.find_by_id(event_id)
        if event is None:
            raise errors.NOT_FOUND
        sectors = await self._sectors.list_by_event(event_id)
        return EventDetail(event=event, sectors=sectors)

File: backends/go/internal/usecase/events.go
package usecase

import (
	"context"

	"github.com/ticketing-labs/backend-go/internal/domain"
)

type EventDetail struct {
	Event   domain.Event
	Sectors []domain.Sector
}

type EventService struct {
	events  EventRepository
	sectors SectorRepository
}

func NewEventService(e EventRepository, s SectorRepository) *EventService {
	return &EventService{events: e, sectors: s}
}

// List returns a page of events and an opaque next cursor. An empty next cursor
// means the caller has reached the end.
func (s *EventService) List(ctx context.Context, cursor string, limit int) ([]domain.Event, string, error) {
	if limit <= 0 || limit > 100 {
		limit = 20
	}
	return s.events.List(ctx, cursor, limit)
}

func (s *EventService) Get(ctx context.Context, id string) (*EventDetail, error) {
	event, err := s.events.FindByID(ctx, id)
	if err != nil {
		return nil, domain.ErrNotFound
	}
	sectors, err := s.sectors.ListByEvent(ctx, id)
	if err != nil {
		return nil, domain.ErrInternal
	}
	return &EventDetail{Event: *event, Sectors: sectors}, nil
}
